test/test_sync_subscriptions.cpp

```diff
@@ -532,6 +532,144 @@ TEST(Sync_SubscriptionStoreSubSetHasTable)
     CHECK(table_set.empty());
 }
 
+TEST(Sync_SubscriptionStoreNotifyAll)
+{
+    SHARED_GROUP_TEST_PATH(sub_store_path)
+    SubscriptionStoreFixture fixture(sub_store_path);
+    auto store = SubscriptionStore::create(fixture.db, [](int64_t) {});
+
+    auto mutex = std::mutex();
+    auto cv = std::condition_variable();
+    size_t hit_count = 0;
+
+    auto error_handler = [&](StatusWith<SubscriptionSet::State> state) {
+        CHECK(!state.is_ok());
+        auto status = state.get_status();
+        if (!status.is_ok()) {
+            CHECK_EQUAL(status.code(), ErrorCodes::OperationAborted);
+            CHECK_EQUAL(status.reason(), "operation aborted");
+            std::lock_guard lk(mutex);
+            hit_count++;
+            cv.notify_one();
+        }
+    };
+
+    auto read_tr = fixture.db->start_read();
+    // We should have no subscriptions yet so this should return false.
+    auto table_set = store->get_tables_for_latest(*read_tr);
+    CHECK(table_set.empty());
+
+    Query query_a(read_tr->get_table(fixture.a_table_key));
+    query_a.equal(fixture.foo_col, StringData("JBR")).greater_equal(fixture.bar_col, int64_t(1));
+    Query query_b(read_tr->get_table(fixture.a_table_key));
+    query_b.equal(fixture.foo_col, "Realm");
+
+    // Create multiple pending subscriptions and notify all of them
+    auto mut_sub_set1 = store->get_latest().make_mutable_copy();
+    mut_sub_set1.insert_or_assign(query_a);
+    auto sub_set1 = mut_sub_set1.commit();
+
+    sub_set1.get_state_change_notification(SubscriptionSet::State::Complete)
+        .get_async([&](StatusWith<SubscriptionSet::State> state) {
+            error_handler(state);
+        });
+
+    auto mut_sub_set2 = store->get_latest().make_mutable_copy();
+    mut_sub_set2.insert_or_assign(query_b);
+    auto sub_set2 = mut_sub_set2.commit();
+
+    sub_set2.get_state_change_notification(SubscriptionSet::State::Complete)
+        .get_async([&](StatusWith<SubscriptionSet::State> state) {
+            error_handler(state);
+        });
+
+    auto mut_sub_set3 = store->get_latest().make_mutable_copy();
+    mut_sub_set3.insert_or_assign(query_a);
+    auto sub_set3 = mut_sub_set3.commit();
+
+    sub_set3.get_state_change_notification(SubscriptionSet::State::Complete)
+        .get_async([&](StatusWith<SubscriptionSet::State> state) {
+            error_handler(state);
+        });
+
+    store->notify_all_state_change_notifications(Status{ErrorCodes::OperationAborted, "operation aborted"});
+    {
+        std::unique_lock lk(mutex);
+        CHECK(cv.wait_for(lk, std::chrono::seconds(15), [&] {
+            return hit_count == 3;
+        }));
+    }
+}
+
+TEST(Sync_SubscriptionStoreTerminate)
+{
+    SHARED_GROUP_TEST_PATH(sub_store_path)
+    SubscriptionStoreFixture fixture(sub_store_path);
+    auto store = SubscriptionStore::create(fixture.db, [](int64_t) {});
+
+    auto mutex = std::mutex();
+    auto cv = std::condition_variable();
+    size_t hit_count = 0;
```

Comment:
same comment as above about these actually not requiring any special thread safety. all these operations are happening on the same thread.